includes/compat/subscriptions/class-wc-payment-gateway-wcpay-subscriptions-compat.php

```diff
@@ -380,4 +384,89 @@ public function render_custom_payment_meta_input( $subscription, $field_id, $fie
 		}
 		echo '</select>';
 	}
+
+	/**
+	 * Add specific data like last 4 digit of wcpay payment gateway
+	 *
+	 * @param string          $old_payment_method_title Payment method title, eg: Credit card.
+	 * @param string          $old_payment_method Payment gateway id.
+	 * @param WC_Subscription $subscription The subscription order.
+	 * @return string
+	 */
+	public function get_specific_old_payment_method_title( $old_payment_method_title, $old_payment_method, $subscription ) {
+		// make sure payment method is wcpay's.
+		if ( WC_Payment_Gateway_WCPay::GATEWAY_ID !== $old_payment_method ) {
+			return $old_payment_method_title;
+		}
+
+		$new_payment_method = $subscription->get_payment_method();
+
+		// if old and new payment method are different, no need to show specific.
+		if ( $old_payment_method !== $new_payment_method ) {
+			return $old_payment_method_title;
+		}
+
+		$last_order_id = $subscription->get_last_order();
+		if ( ! $last_order_id ) {
+			return $old_payment_method_title;
+		}
+
+		$last_order = wc_get_order( $last_order_id );
+		$token_ids  = $last_order->get_payment_tokens();
+		// since old payment must be the second to last saved payment...
+		if ( count( $token_ids ) < 2 ) {
+			return $old_payment_method_title;
+		}
+
+		$second_to_last_token_id = $token_ids[ count( $token_ids ) - 2 ];
+		$token                   = WC_Payment_Tokens::get( $second_to_last_token_id );
+		if ( ! $token || ! ( $token instanceof WC_Payment_Token_CC && method_exists( $token, 'get_last4' ) ) ) {
+			return $old_payment_method_title;
+		}
+
+		// translators: 1: payment method likely credit card, 2: last 4 digit.
+		return sprintf( __( '%1$s ending in %2$s', 'woocommerce-payments' ), $old_payment_method_title, $token->get_last4() );
+	}
+
+	/**
+	 * Add specific data like last 4 digit of wcpay payment gateway
+	 *
+	 * @param string          $new_payment_method_title Payment method title, eg: Credit card.
+	 * @param string          $new_payment_method Payment gateway id.
+	 * @param WC_Subscription $subscription The subscription order.
+	 * @return string
+	 */
+	public function get_specific_new_payment_method_title( $new_payment_method_title, $new_payment_method, $subscription ) {
+		// make sure payment method is wcpay's.
+		if ( WC_Payment_Gateway_WCPay::GATEWAY_ID !== $new_payment_method ) {
+			return $new_payment_method_title;
+		}
+
+		$old_payment_method = $subscription->get_meta( '_old_payment_method' );
+
+		// if old and new payment method are different, no need to show specific.
+		if ( $new_payment_method !== $old_payment_method ) {
+			return $new_payment_method_title;
+		}
+
+		$request = isset( $_POST ) ? $_POST : []; // phpcs:ignore WordPress.Security.NonceVerification.Missing
+		$token   = WCPay\Payment_Information::get_token_from_request( $request );
+		if ( $token && $token instanceof WC_Payment_Token_CC && method_exists( $token, 'get_last4' ) ) {
```

Comment:
Same as above regarding `method_exists`. 
That line alse makes me feel we need to extract title composition from token to own method.